File: win-prog/ProgWinVB/Buttons and Labels and Scrolls (Oh, My!)/CheckBoxDemo/CheckBoxDemo.vb
'---------------------------------------------
' CheckBoxDemo.vb (c) 2002 by Charles Petzold
'---------------------------------------------
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Class CheckBoxDemo
    Inherits Form

    Shared Sub Main()
        Application.Run(New CheckBoxDemo())
    End Sub

    Sub New()
        Text = "CheckBox Demo"

        Dim achkbox(3) As CheckBox
        Dim cyText As Integer = Font.Height
        Dim cxText As Integer = cyText \ 2
        Dim astrText() As String = {"Bold", "Italic", _
                                    "Underline", "Strikeout"}
        Dim i As Integer

        For i = 0 To 3
            achkbox(i) = New CheckBox()
            achkbox(i).Text = astrText(i)
            achkbox(i).Location = New Point(2 * cxText, _
                                           (4 + 3 * i) * cyText \ 2)
            achkbox(i).Size = New Size(12 * cxText, cyText)
            AddHandler achkbox(i).CheckedChanged, _
                                        AddressOf CheckBoxOnCheckedChanged
        Next i
        Controls.AddRange(achkbox)
    End Sub

    Private Sub CheckBoxOnCheckedChanged(ByVal obj As Object, _
                                         ByVal ea As EventArgs)
        Invalidate(False)
    End Sub

    Protected Overrides Sub OnPaint(ByVal pea As PaintEventArgs)
        Dim grfx As Graphics = pea.Graphics
        Dim fs As FontStyle = 0
        Dim afs() As FontStyle = {FontStyle.Bold, FontStyle.Italic, _
                                  FontStyle.Underline, FontStyle.Strikeout}
        Dim i As Integer

        For i = 0 To 3
            If DirectCast(Controls(i), CheckBox).Checked Then
                fs = fs Or afs(i)
            End If
        Next i

        Dim fnt As New Font(Font, fs)
        grfx.DrawString(Text, fnt, New SolidBrush(ForeColor), 0, 0)
    End Sub
End Class

File: win-prog/ProgWinCS/Buttons and Labels and Scrolls (Oh, My!)/CheckBoxDemo/CheckBoxDemo.cs
//-------------------------------------------
// CheckBoxDemo.cs © 2001 by Charles Petzold
//-------------------------------------------
using System;
using System.Drawing;
using System.Windows.Forms;

class CheckBoxDemo: Form
{
     public static void Main()
     {
          Application.Run(new CheckBoxDemo());
     }
     public CheckBoxDemo()
     {
          Text = "CheckBox Demo";

          CheckBox[] achkbox  = new CheckBox[4];
          int        cyText   = Font.Height;
          int        cxText   = cyText / 2;
          string[]   astrText = {"Bold", "Italic", "Underline", "Strikeout"};

          for (int i = 0; i < 4; i++)
          {
               achkbox[i] = new CheckBox();
               achkbox[i].Text = astrText[i];
               achkbox[i].Location = new Point(2 * cxText, 
                                               (4 + 3 * i) * cyText / 2);
               achkbox[i].Size = new Size(12 * cxText, cyText);
               achkbox[i].CheckedChanged += 
                              new EventHandler(CheckBoxOnCheckedChanged);
          }
          Controls.AddRange(achkbox);
     }
     void CheckBoxOnCheckedChanged(object obj, EventArgs ea)
     {
          Invalidate(false);
     }
     protected override void OnPaint(PaintEventArgs pea)
     {
          Graphics    grfx = pea.Graphics;
          FontStyle   fs   = 0;
          FontStyle[] afs  = { FontStyle.Bold,      FontStyle.Italic, 
                               FontStyle.Underline, FontStyle.Strikeout };

          for (int i = 0; i < 4; i++)
               if (((CheckBox) Controls[i]).Checked)
                    fs |= afs[i];

          Font font = new Font(Font, fs);
          grfx.DrawString(Text, font, new SolidBrush(ForeColor), 0, 0);
     }
}
